File: src/model/IRectShape.cpp
/*

  INScore Project

  Copyright (C) 2009,2010  Grame

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA

  Grame Research Laboratory, 9 rue du Garet, 69001 Lyon - France
  research@grame.fr

*/

#include <iostream>

#include "IRectShape.h"
#include "IMessage.h"
#include "Updater.h"

using namespace std;

namespace inscore
{

const string IRectShape::kRectShapeType("rshape");

//--------------------------------------------------------------------------
IRectShape::IRectShape( const std::string& name, IObject* parent ) : IShapeMap(name, parent)
{ 
	fTypeString = kRectShapeType;
	fGetMsgHandlerMap[""]					= TGetParamMethodHandler<IRectShape, TFloatSize (IRectShape::*)() const>::create(this, &IRectShape::getDimension);
	fMsgHandlerMap[kwidth_GetSetMethod]		= TSetMethodMsgHandler<IObject, float>::create(this, &IObject::setWidth);
	fMsgHandlerMap[kheight_GetSetMethod]	= TSetMethodMsgHandler<IObject, float>::create(this, &IObject::setHeight);
    
	fSigHandlerMap[kwidth_GetSetMethod]		= TSetMethodSigHandler<IObject, float>::create(this, &IObject::setWidth);
	fSigHandlerMap[kheight_GetSetMethod]	= TSetMethodSigHandler<IObject, float>::create(this, &IObject::setHeight);
}

//--------------------------------------------------------------------------
void IRectShape::accept (Updater* u)
{
	u->updateTo (SIRectShape(this));
}

//--------------------------------------------------------------------------
MsgHandler::msgStatus IRectShape::set (const IMessage* msg)	
{
	MsgHandler::msgStatus status = IObject::set(msg);
	if (status & (MsgHandler::kProcessed + MsgHandler::kProcessedNoChange)) return status; 
	
	if (msg->size() == 3) {
		const float err = -1;
		float width = msg->param(1)->value<float>(err);
		float height = msg->param(2)->value<float>(err);

		if ( (width != err) && (height != err) && ( ( width != getWidth() ) || ( height != getHeight() ) ) )
		{
			setWidth( width );
			setHeight( height );
			newData(true);
			status = MsgHandler::kProcessed;
		}
		else status = MsgHandler::kProcessedNoChange;
	}
	else status = MsgHandler::kBadParameters;
	return status;
}
		
}

File: src/javascript/src/model/IRectShape.ts
///<reference path="../lib/TEnums.ts"/>
///<reference path="IObject.ts"/>

class IRectShape extends IObject {

    fBrushStyle : brushStyle;
    protected fModified : boolean;

    fBrushStyleStr2Num: { [id: string] : brushStyle; } = {};
    fBrushStyleNum2Str: { [id: number] : string; } = {};


    constructor(name: string, parent: IObject) {
        super(name, parent);
        this.fBrushStyle = brushStyle.solid;
        this.buildBrushStyle();
    }

    buildBrushStyle(): void {
        this.fBrushStyleStr2Num["solid"] = brushStyle.solid;
        this.fBrushStyleStr2Num["none"]  = brushStyle.none;

        this.fBrushStyleNum2Str[brushStyle.solid] = "solid";
        this.fBrushStyleNum2Str[brushStyle.none]  = "none";
    }

    create(name: string, parent: IObject): IRectShape { return new IRectShape(name, parent); }

    setHandlers() {
    	super.setHandlers(); 
        this.fMsgHandlerMap[kwidth_GetSetMethod]  = new TMsgHandlerNum(this._setWidth());
        this.fMsgHandlerMap[kheight_GetSetMethod] = new TMsgHandlerNum(this._setHeight());

        this.fGetMsgHandlerMap[kbrushStyle_GetSetMethod] = new TGetMsgHandlerText(this._getBrushStyle());
        this.fMsgHandlerMap[kbrushStyle_GetSetMethod]    = new TMsgHandlerText(this._setBrushStyle());
    }
   
    set(msg: IMessage): msgStatus {
        let status = super.set(msg);
       
        // Cas ou le type est différent, le proxy est utilisé dans super.set()
        if (status & (msgStatus.kProcessed + msgStatus.kProcessedNoChange)) return status; 

        // Vérification des paramêtres du message "set"
        if (msg.size() == 4) {
            let width = msg.paramNum(2), height = msg.paramNum(3);
            if (!width.correct || !height.correct) { return msgStatus.kBadParameters; }

            // Si les valeurs des attributs ne sont pas les mêmes
            if ( (width.value != this.fPosition.getWidth()) || (height.value != this.fPosition.getHeight()) ) {
                this.fPosition.setWidth(width.value);
                this.fPosition.setHeight(height.value);
                this.newData(true);
                status = msgStatus.kProcessed;
            }
            else status = msgStatus.kProcessedNoChange;
        }
        else status = msgStatus.kBadParameters;
        return status;
    }  

    getSet(): IMessage	{ 
    	return new IMessage(this.getOSCAddress(), [kset_SetMethod, this.fTypeString, this.fPosition.getWidth(), this.fPosition.getHeight()]); 
    }
   
    //accept (Updater*): void
    // virtual MsgHandler::msgStatus set (const IMessage* msg);

    getBrushStyle () : string {return this.fBrushStyleNum2Str[this.fBrushStyle];}
    _getBrushStyle() : GetStringMethod { return () => this.getBrushStyle() }
    setBrushStyle (brushStyle : string): void {
        let style = this.fBrushStyleStr2Num[brushStyle];
        this.fBrushStyle = style;}
    _setBrushStyle(): SetStringMethod { return (brush : string) => this.setBrushStyle(brush) }
}